webapp/channels/src/components/suggestion/at_mention_provider/at_mention_provider.tsx

```diff
@@ -1,25 +1,73 @@
 // See LICENSE.txt for license information.
 
-import {getSuggestionsSplitBy, getSuggestionsSplitByMultiple} from 'mattermost-redux/utils/user_utils';
-import {makeGetProfilesInChannel} from 'mattermost-redux/selectors/entities/users';
-import {makeAddLastViewAtToProfiles} from 'mattermost-redux/selectors/entities/utils';
+import { ActionResult } from 'mattermost-redux/types/actions';
+import { getSuggestionsSplitBy, getSuggestionsSplitByMultiple } from 'mattermost-redux/utils/user_utils';
+import {Group} from '@mattermost/types/groups';
+import { makeGetProfilesInChannel } from 'mattermost-redux/selectors/entities/users';
+import { makeAddLastViewAtToProfiles } from 'mattermost-redux/selectors/entities/utils';
+import {UserProfile, UserProfileWithLastViewAt} from '@mattermost/types/users';
 
 import store from 'stores/redux_store';
-import {Constants} from 'utils/constants';
+import { Constants } from 'utils/constants';
+import { GlobalState } from 'types/store';
 
 import Provider from '../provider';
 
 import AtMentionSuggestion from './at_mention_suggestion';
 
-const profilesInChannelOptions = {active: true};
+const profilesInChannelOptions = { active: true };
 const regexForAtMention = /(?:^|\W)@([\p{L}\d\-_. ]*)$/iu;
 
+type Filters = {
+    role?: string;
+    inactive?: boolean;
+    active?: boolean;
+    roles?: string[];
+    exclude_roles?: string[];
+    channel_roles?: string[];
+    team_roles?: string[];
+};
+
+type Props = {
+    currentUserId:string; 
+    channelId:string; 
+    autocompleteUsersInChannel:(prefix: string, channelId: string) => Promise<ActionResult>; 
+    useChannelMentions:boolean; 
+    autocompleteGroups:Array<{ id: string }> | null; 
+    searchAssociatedGroupsForReference(prefix: string, teamId: string, channelId: string | undefined) => Promise<{ data: any }>; 
+    priorityProfiles:UserProfile[];
+}
+
+type Results = {
+    matchedPretext: string;
+    terms: string[];
+    items: any;
+    component: React.ElementType;
+}
+
+type ResultsCallback = (results: Results) => void;
+
 // The AtMentionProvider provides matches for at mentions, including @here, @channel, @all,
 // users in the channel and users not in the channel. It mixes together results from the local
 // store with results fetched from the server.
 export default class AtMentionProvider extends Provider {
-    constructor(props) {
+    data: null;
+    lastCompletedWord: string;
+    lastPrefixWithNoResults: string;
+    triggerCharacter: string;
+    getProfilesInChannel: (state: GlobalState, channelId: string, filters?: Filters | undefined) => UserProfile[];
+    addLastViewAtToProfiles: (state: GlobalState, profiles: UserProfile[]) => UserProfileWithLastViewAt[];
+
+    currentUserId: any;
+    channelId: any;
+    autocompleteUsersInChannel: any;
+    useChannelMentions: any;
+    autocompleteGroups: any;
+    searchAssociatedGroupsForReference: any;
+    priorityProfiles: any;
```

Comment:
I don't know if this is correct. Please verify this.

Reply:
Mattermost is very strict about typings. Please avoid using `any` and use the correct types. For instance `currentUserId` and `channelId` should be `string`

Reply:
Yeah, initially I assigned the respective types only. But similar types of errors were showing up for each variable that I haven't assigned type: `any`.
```
Property 'currentUserId' has no initializer and is not definitely assigned in the constructor.
```

Reply:
Yes, as the error said, you need to initialize `currentUserId` in the constructor to empty string.
Or alternatively, you can change the type to `currentUserId?: string;`. The question mark indicates that `currentUserId ` can be undefined so you dont need to initialize it. 

However, I personally prefer the first option.

Reply:
I have initialized the types in the constructor. But I couldn't understand what `autocompleteUsersInChannel`, `autocompleteGroups`, `searchAssociatedGroupsForReference`, and `priorityProfiles` should be defined as. So I initialized them with a `?`.

Reply:
One trick you can do is to check where `AtMentionProvider` is used. If it is used in another typescript file, you can see the types passed. For example I see that it used inside `Textbox`, so you can go and check the types passed to `AtMentionProvider `

https://github.com/mattermost/mattermost/blob/a1956948961d4e86ea799edc56a723a8aebc8752/webapp/channels/src/components/textbox/textbox.tsx#L108-L116